CitiesSkylinesMods/TMPE
# Extend `CheckboxOption` with new properties

Fixes: #1299
Updates: #1279
Prerequisite for #1267

- Add `ReadOnlyUI` property
    - When `true`, checkbox is readonly and visually dimmed
- Add `Scope` property and `PersistTo` flags enum
    - Indicates where the value can be persisted (Global, Savegame, etc)
- Add `IsInScope` property
    - When `false` the UI component should be made read-only
    - In this PR, I've commented out the code that does that
    - ...it will be revived when #1356 OP is complete
- Add `Translator` property
    - Allows use of alternate translation lookups (`Translation.Options` is default)
- Add `HasUI` property
    - Becomes `true` after `AddUI()` has been used
- Add `PropagatesTrueTo` property (optional)
    - If specified, setting checkbox `true` will propagate that to listed checkboxes
    - See `OptionsGeneralTab_DebugCheckbox.cs` for how to use it
- Add `PropagatesFalseTo` property
    - If specified, setting checkbox `false` will propagate that to listed checkboxes
    - This is automatically managed by the `PropagatesTrueTo` property of other checkboxes
- Remove requirement for `Options` shims
    - Checkbox will use internal data store when `PersistTo.Savegame` flag is not set
- Add logging when checkbox options change value
    - Should help with future support
    - May also help identify if we get any recursion or other excessive amounts of toggles
    - Logging will be reduced quite a bit when #1356 OP is complete

![image](https://user-images.githubusercontent.com/1386719/150624230-7118f5d0-fc1b-4be2-bc5f-ae566402e225.png)
